File: src/main/java/com/plantcare/plantcareassistant/entities/Plant.java
package com.plantcare.plantcareassistant.entities;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.JsonNode;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;

@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public class Plant {
    private String id;

    @JsonProperty("common_name")
    private String commonName;

    @JsonProperty("scientific_name")
    private String scientificName;


    private String type;
    private boolean indoor;

    private String watering;

    @JsonProperty("default_image")
    private DefaultImage defaultImage;

    @JsonProperty("care_level")
    private String careLevel;
    private String description;
    private Set<String> sunlight;

    //json node can handle both arrays and objects
    //need this because data is stored inconsistently in the api
    @JsonProperty("depth_water_requirement")
    private JsonNode depthWaterRequirement;
    @JsonProperty("watering_general_benchmark")
    private WateringTime wateringTime;

    private String imageUrl;


    public List<WaterRequirement> getDepthWaterRequirements() {
        List<WaterRequirement> requirements = new ArrayList<>();
        if (depthWaterRequirement.isArray()) {
            // Process as an array
            for (JsonNode node : depthWaterRequirement) {
                WaterRequirement req =jsonNodeToWaterRequirement(node);
                        requirements.add(req);
            }
        } else {
            WaterRequirement req = jsonNodeToWaterRequirement(depthWaterRequirement);
                    requirements.add(req);
        }
        return requirements;
    }

    private WaterRequirement jsonNodeToWaterRequirement(JsonNode jsonNode) {
        String unit = jsonNode.has("unit") ? jsonNode.get("unit").asText() : null;
        int value = jsonNode.has("value") ? jsonNode.get("value").asInt() : 0;
        return new WaterRequirement(unit, value);
    }


    public String getImageUrl() {
        if (defaultImage != null) {
            return defaultImage.getMediumUrl();
        }
        return null;
    }

    public void setImageFromDefault(DefaultImage defaultImage) {
        this.imageUrl = defaultImage.getMediumUrl();
    }



    @Getter
    @Setter
    @NoArgsConstructor
    @AllArgsConstructor
    public static class WaterRequirement { //how much water
        private String unit;
        private int value;
    }

    @Getter
    @Setter
    @NoArgsConstructor
    @AllArgsConstructor
    public static class WateringTime { //when to water
        private String value;  // num of days
        private String unit; //hours, weeks, days, etc

    }

    @Getter
    @Setter
    @NoArgsConstructor
    @AllArgsConstructor
    public static class DefaultImage {
        @JsonProperty("medium_url")
        private String mediumUrl;

        public String getMediumUrl() {
            return mediumUrl;
        }

        public void setMediumUrl(String mediumUrl) {
            this.mediumUrl = mediumUrl;
        }
    }


}

File: AndroidApp/app/src/main/java/com/example/plantcare/data/model/Plant.kt
package com.example.plantcare.data.model

import com.google.gson.annotations.SerializedName

import com.google.gson.*
import com.google.gson.annotations.JsonAdapter
import java.lang.reflect.Type

data class Plant(
    val id: String,
    @SerializedName("common_name") val commonName: String,
    @SerializedName("scientific_name") val scientificName: List<String>,
    val type: String,
    val indoor: Boolean,
    val watering: String,
    @SerializedName("default_image") val defaultImage: DefaultImage,
    @SerializedName("care_level") val careLevel: String,
    val description: String,
    val sunlight: Set<String>,
    @JsonAdapter(DepthWaterRequirementDeserializer::class)
    @SerializedName("depth_water_requirement") val depthWaterRequirement: List<WaterRequirement>?,
    @SerializedName("watering_general_benchmark") val wateringTime: WateringTime?,
    val imageUrl: String
)

data class WaterRequirement(
    val unit: String?,
    val value: Int
)

data class WateringTime(
    val value: String,
    val unit: String
)

data class DefaultImage(
    @SerializedName("medium_url") val mediumUrl: String
)

class DepthWaterRequirementDeserializer : JsonDeserializer<List<WaterRequirement>> {
    override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): List<WaterRequirement> {
        val result = mutableListOf<WaterRequirement>()
        if (json.isJsonArray) {
            val array = json.asJsonArray
            for (element in array) {
                result.add(context.deserialize(element, WaterRequirement::class.java))
            }
        } else if (json.isJsonObject) {
            result.add(context.deserialize(json, WaterRequirement::class.java))
        }
        return result
    }
}
